apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex

```diff
@@ -0,0 +1,78 @@
+defmodule Lexical.RemoteControl.CodeMod.Rename.Prepare do
+  alias Lexical.Ast
+  alias Lexical.Ast.Analysis
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Lexical.Formats
+  alias Lexical.RemoteControl.CodeIntelligence.Entity
+  alias Lexical.RemoteControl.CodeMod.Rename
+  alias Lexical.RemoteControl.Search.Store
+
+  require Logger
+
+  @spec prepare(Analysis.t(), Position.t()) ::
+          {:ok, {atom(), String.t()}, Range.t()} | {:error, term()}
+  def prepare(%Analysis{} = analysis, %Position{} = position) do
+    case resolve(analysis, position) do
+      {:ok, {:module, module}, range} ->
+        {:ok, Formats.module(module), range}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @spec resolve(Analysis.t(), Position.t()) ::
+          {:ok, {atom(), atom()} | {atom(), tuple()}, Range.t()} | {:error, term()}
+  def resolve(%Analysis{} = analysis, %Position{} = position) do
+    case do_resolve(analysis, position) do
+      {:ok, {:module, _module}, _range} ->
+        {module, range} = surround_the_whole_module(analysis, position)
+
+        if cursor_at_declaration?(module, range) do
+          {:ok, {:module, module}, range}
+        else
+          {:error, {:unsupported_location, :module}}
+        end
+
+      other ->
+        other
+    end
+  end
+
+  defp surround_the_whole_module(analysis, position) do
+    # When renaming occurs, we want users to be able to choose any place in the defining module,
+    # not just the last local module, like: `defmodule |Foo.Bar do` also works.
+    {:ok, %{end: {_end_line, end_character}}} = Ast.surround_context(analysis, position)
+    end_position = %{position | character: end_character - 1}
+    {:ok, {:module, module}, range} = do_resolve(analysis, end_position)
+    {module, range}
+  end
+
+  defp cursor_at_declaration?(module, rename_range) do
+    case Store.exact(module, type: :module, subtype: :definition) do
+      {:ok, [definition]} ->
+        rename_range == definition.range
```

Comment:
This method is reliable because it compares the range of the module after 'surround' with the defined range of 'exact'. However, do you have any better suggestions for the comparison?

Reply:
struct comparison does a deep compare, there are fields in the `Position` struct that shouldn't have a bearing on them being equal. I think you want to compare ranges based on their positions' lines and characters.

Reply:
@scohen So should we implement an equal? function for range or position to do this?

Reply:
@scottming #767 implements this exact function but is currently dependent on #768. It's only dependent, however, because #767 refactors `Range.contains?/2` to be in terms of `Position.compare/2`, which causes tests to fail without #768.

I can very easily move things around such that #767 can be merged in first without #768, then you can rebase and use that function here.

Reply:
#767 has been updated so that it can be merged independently if someone is able to review it.

Reply:
@zachallaun thanks